### PR TITLE
Add function artifact check in deployFunction

### LINKED ISSUE #3580: serverless deploy function ignores package artifact
# This is a Bug Report

### What went wrong?
Consider the below serverless.yml file
```
service: test-service

provider:
  name: aws
  runtime: python2.7

functions:
  hello:
    handler: handler.hello
    package:
      artifact: upload.zip
```

When `serverless deploy function -f hello` is run, Serverless ignores the `functions.hello.package.artifact` configuration and falls back to whatever exclude/include configuration it determines.
This indicates that `serverless deploy function -f` is currently ignoring the `functions.[function name].package.artifact` config  
### What did you expect should have happened?
I expected `serverless deploy function -f` to respect the `functions.[function name].package.artifact` configuration the same way `serverless deploy` does.
 
### What was the config you used?
See description

### What stacktrace or error message from your provider did you see?
No error but the behavior is inconsistent with expectations

## Additional Data

* ***Serverless Framework Version you're using***: 1.12.1
* ***Operating System***: OS X El Capitan
* ***Stack Trace***:
* ***Provider Error messages***:

### LINKED ISSUE #3625: serverless deploy function -f not packaging correctly.
I'm having issues when running.

`serverless deploy function -f  <function>` 

I haven't determined what version this was introduced on, but if I run a serverless deploy everything seems to be deployed correctly. If I deploy a single function I received the following error.

```
{
  "errorMessage": "Class not found: com.xxxxxxx.user.RequestHandler",
  "errorType": "class java.lang.ClassNotFoundException"
}
```

This is a Maven project running Java 1.8 and Serverless 1.13.2. 

On thing that I notice is the package size is very different between Serverless 1.10 and Serverless 1.13.2. It goes from 24.27mb -> 30.25mb

Also in my serverless.yml file I'm explicitly referencing the the package artifact for every function. This was needed awhile back to get the individual functions to deploy with a maven package. Is it maybe not needed any longer and that is my underlying issue? 

```
package:
      artifact: target/xxxxxxxx-${self:custom.stage}.jar
```

### PR DESCRIPTION
**Pull request by darkowlzz**
## What did you implement:

Closes #3625 
Closes #3580 

## How did you implement it:

Added function artifact check in `deployFunction()` and skip packing when an artifact is provided.
And add function artifact check in `packageService()` to use the provided artifact path and not derive some default artifact path.

## How can we verify it:

* Create an aws-nodejs template project and deploy.
* Create a dummy zip file, say `foo.zip`, and add that as artifact of a function.
```
functions:
  hello:
    handler: handler.hello
    package:
      artifact: foo.zip
```

* Run `sls deploy function -f hello` to deploy the function.
* Open AWS lambda console and export the function. The downloaded file is the same dummy zip file.

## Todos:

- [x] Write tests
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/plugins/aws/deployFunction/index.js [hunk 1]
```diff
@@ -66,9 +66,14 @@ class AwsDeployFunction {
   }
 
   deployFunction() {
-    const artifactFileName = this.provider.naming
-      .getFunctionArtifactName(this.options.function);
-    const artifactFilePath = path.join(this.packagePath, artifactFileName);
+    const functionObject = this.serverless.service.getFunction(this.options.function);
+    let artifactFilePath = functionObject.artifact;
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.js]: We should add a test for that

File: lib/plugins/package/lib/packageService.js [hunk 2]
```diff
@@ -70,12 +71,21 @@ module.exports = {
     const functionObject = this.serverless.service.getFunction(functionName);
     const funcPackageConfig = functionObject.package || {};
 
+    // use the artifact in function config if provided
+    if (funcPackageConfig.artifact) {
+      const filePath = path.join(this.serverless.config.servicePath, funcPackageConfig.artifact);
+      functionObject.package.artifact = filePath;
```
Comment (darkowlzz) [inline lib/plugins/package/lib/packageService.js]: I found out that this should be `functionObject.package.artifact` and not `functionObject.artifact`. I referred to the code below earlier. Changed that as well.

File: lib/plugins/aws/deployFunction/index.test.js [hunk 3]
```diff
@@ -139,5 +139,24 @@ describe('AwsDeployFunction', () => {
...
+        handler: true,
+        package: {
+          artifact: 'foo.zip',
+        },
+      });
+      sinon.stub(awsDeployFunction.provider, 'request').resolves();
+
+      return awsDeployFunction.deployFunction().then(() => {
+        // const expected = 'foo.zip';
+        // expect(fs.readFileSync.calledWithExactly(expected)).to.be.equal(true);
```
Comment (darkowlzz) [inline lib/plugins/aws/deployFunction/index.test.js]: This test checks when `package.artifact` is provided, it's read by `fs.readFileSync()`.

File: lib/plugins/aws/deployFunction/index.js [hunk 4]
```diff
@@ -66,9 +66,15 @@ class AwsDeployFunction {
   }
 
   deployFunction() {
-    const artifactFileName = this.provider.naming
-      .getFunctionArtifactName(this.options.function);
-    const artifactFilePath = path.join(this.packagePath, artifactFileName);
+    const functionObject = this.serverless.service.getFunction(this.options.function);
+    const funcPackageConfig = functionObject.package || {};
+    let artifactFilePath = funcPackageConfig.artifact;
```
Comment (darkowlzz) [inline lib/plugins/aws/deployFunction/index.js]: @pmuens  I need help here. Whenever the `artifact` is read, I get some parsing errors and I'm not sure why that's happening. The error messages don't seem to have anything related to this part of the code. Can you run and have a look?

Comment (darkowlzz) [inline lib/plugins/aws/deployFunction/index.js]: The error showed up in travis as well.
```
Failed to parse file: /home/travis/build/serverless/serverless/lib/plugins/aws/invokeLocal/fixture/handlerWithError.js
Transformation error; return original code
{ [Error: Line 1: Unexpected identifier]
  lineNumber: 1,
  description: 'Unexpected identifier',
  index: 20 }
```
Not sure why the build exited with 0. 🤷‍♀️

File: lib/plugins/aws/deployFunction/index.test.js [hunk 5]
```diff
@@ -139,5 +139,24 @@ describe('AwsDeployFunction', () => {
         fs.statSync.restore();
       });
     });
+
+    it('should read the provided artifact', () => {
+      sinon.stub(fs, 'readFileSync').returns('some dummy data');
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.test.js]: All the stubs should be moved in a `beforeEach()` block and reset in an `afterEach()` block.

File: lib/plugins/aws/deployFunction/index.test.js [hunk 6]
```diff
@@ -139,5 +139,24 @@ describe('AwsDeployFunction', () => {
...
+        package: {
+          artifact: 'foo.zip',
+        },
+      });
+      sinon.stub(awsDeployFunction.provider, 'request').resolves();
+
+      return awsDeployFunction.deployFunction().then(() => {
+        // const expected = 'foo.zip';
+        // expect(fs.readFileSync.calledWithExactly(expected)).to.be.equal(true);
+        awsDeployFunction.provider.request.restore();
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.test.js]: The restoring should happen in an `afterEach()` block.

**Comment by darkowlzz** [comment]
@pmuens thank you. You were right. Putting the stubbing and restoring into beforeEach and afterEach solved the issue 🙌

**Comment by pmuens** [comment]
> @pmuens thank you. You were right. Putting the stubbing and restoring into beforeEach and afterEach solved the issue 🙌

Awesome! Thanks for updating @darkowlzz 👍 

Next up we'll take this for another spin.

File: lib/plugins/aws/deployFunction/index.js [hunk 7]
```diff
@@ -66,9 +66,15 @@ class AwsDeployFunction {
...
+    // if function artifact is not provided, derive the default artifact path
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.js]: I would suggest to switch this statement and assume that the user is not using an artifact at the function package level (since most of the time this would be the case).

File: lib/plugins/aws/deployFunction/index.test.js [hunk 8]
```diff
@@ -139,5 +139,33 @@ describe('AwsDeployFunction', () => {
         fs.statSync.restore();
       });
     });
+
+    describe('when artifact is provided', () => {
+      let fsReadFileStub;
+
+      beforeEach(() => {
+        fsReadFileStub = sinon.stub(fs, 'readFileSync').returns('some dummy data');
+        sinon.stub(fs, 'statSync').returns({ size: 2048 });
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.test.js]: We should add an `expect` that this stub was also called.

File: lib/plugins/aws/deployFunction/index.test.js [hunk 9]
```diff
@@ -139,5 +139,33 @@ describe('AwsDeployFunction', () => {
...
+        sinon.stub(serverless.service, 'getFunction').returns({
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.test.js]: We should add an `expect` that this stub was also called.

File: lib/plugins/aws/deployFunction/index.test.js [hunk 10]
```diff
@@ -139,5 +139,33 @@ describe('AwsDeployFunction', () => {
...
+          handler: true,
+          package: {
+            artifact: 'foo.zip',
+          },
+        });
+        sinon.stub(awsDeployFunction.provider, 'request').resolves();
```
Comment (pmuens) [inline lib/plugins/aws/deployFunction/index.test.js]: We should add an `expect` that this stub was also called.

File: lib/plugins/package/lib/packageService.js [hunk 11]
```diff
@@ -70,12 +71,21 @@ module.exports = {
...
+      return BbPromise.resolve(filePath);
+    }
+
     const exclude = this.getExcludes(funcPackageConfig.exclude);
     const include = this.getIncludes(funcPackageConfig.include);
     const zipFileName = `${functionName}.zip`;
 
     return this.zipService(exclude, include, zipFileName).then(artifactPath => {
-      functionObject.artifact = artifactPath;
```
Comment (pmuens) [inline lib/plugins/package/lib/packageService.js]: Could you give a little  more context why this change is needed?

Comment (darkowlzz) [inline lib/plugins/package/lib/packageService.js]: I found that function object returned by `serverless.service.getFunction()` has `artifact` inside `package`, similar to how we specify them in `serverless.yaml`,
```
functions:
  hello:
    handler: handler.hello
    package:
      artifact: artifactName.zip
```
Along with `exclude` and `include`. Is that not the case?

**Comment by darkowlzz** [comment]
> Unfortunately I was able to break it with the following serverless.yml config when running serverless deploy (which works in master):

@pmuens that was happening because of how the `extendedValidate()` was getting artifact name and path. Changed that to check function object for artifact. Seems to work now. 🙂

**Comment by eahefnawy** [comment]
@darkowlzz @pmuens thanks for that. I've also took a deep dive and test it with the following edge cases, with one unfortunate failed test:

- ✅  `sls deploy` with packaging individually
- ✅  `sls deploy` with packaging with package path
- ✅  `sls deploy` with packaging with artifact at function level
- ✅  `sls deploy` with packaging with artifact at service level
- ✅  `sls deploy function` with packaging individually
- ✅  `sls deploy functon` with packaging with package path
- ✅  `sls deploy function` with packaging with artifact at function level
- ❌  `sls deploy function` with packaging with artifact at service level.

@pmuens could you try to reproduce? with the following config:

```
package:
  artifact: dummy.zip # just zip any file you have on your machine and put it in your service

functions:
  hello:
    handler: handler.hello
```
- run `sls deploy`, 
- go to AWS and export/download the hello function zip and inspect the contents, you'll find that the content you zipped exists rather than the default service contents, which is what is desired.
- run `sls deploy function --hello`
- go to AWS and export/download the hello function zip and inspect the contents, you'll find that the content you zipped is actually the service default, which is not cool because we've specified a zip file at the service level to be used by all functions without an explicit artifact

**Comment by pmuens** [comment]
@eahefnawy thanks for the in-depth test and the list with the different cases you've tested against.

I was able to reproduce the issue. However this issue is also in `master` so it seems like it's unrelated to this PR.

@eahefnawy could you try to reproduce the issue in `master` and confirm that it's not related to this PR?

**Comment by eahefnawy** [comment]
@darkowlzz @pmuens I just fixed this issue and opened another follow up PR based on this PR. Thanks a lot for your efforts here @darkowlzz, they're all still included in the follow up PR 😊

**Comment by pmuens** [comment]
> @darkowlzz @pmuens I just fixed this issue and opened another follow up PR based on this PR. Thanks a lot for your efforts here @darkowlzz, they're all still included in the follow up PR 😊

Great! Thanks for that @eahefnawy 👍 

Should we close this one?

**Comment by eahefnawy** [comment]
@pmuens since all the commits here are included in my PR, I think Github will mark this PR as merged too once my PR is merged. So let's just leave it open and let Github do the git magic ✨

**Comment by pmuens** [comment]
> @pmuens since all the commits here are included in my PR, I think Github will mark this PR as merged too once my PR is merged. So let's just leave it open and let Github do the git magic ✨

Alright. That makes sense 👍
